### DISCUSSION TITLE
Performance issue?

### DISCUSSION THREAD

**Discussion by thany** [streetcomplete/StreetComplete#4135] [Q&A]
I'm wondering if this is an issue with the app, or with my phone. I'm on a Galaxy S9, which sure is an oldie by modern standards, but it works for me. The phone gets REALLY hot while using the app, even when I'm not in direct sunlight. The battery drains quite quickly. I know that for a half hour session, my battery better be above 60% or I won't make it before super battery saver mode kicks in at 25%.

So I wonder, apart from the developer improving CPU usage of the app (if that's still even possible), is there anything I can try to make it reduce its load on the CPU and battery? JSYK, I'm not buying a new phone 🙂

**Comment by matkoniecz** [comment]
You can try uninstalling and installing app again - maybe something got corrupted, got into bizarre interaction?

For less radical but similar - flushing cache on system or app level in settings.

In both cases: make sure that your changes were uploaded (you can check it on profile page with statistics).

If you can - you can try profiling app and report which part is unusually battery consuming.

> if that's still even possible

See for example #4078 #4077 #4079 though I guess that this is unlikely to help significantly with battery. And compass draw update is not very smart right now and it may be using noticeable amount of battery.

**Reply by thany** [reply]
Thanks. I've deleted the cache in both places. I'll probably have another walk later this evening or tomorrow, so let's see how this goes. Without (much) improvement I'll do a reinstall.

In the mean time, I've never profiled an app. So I'm guessing that's not easy to do?

**Reply by matkoniecz** [reply]
I am not promising that reinstall will help: but at least in some cases it may help (I used it with SC before and succeeded when something went wrong for unclear reason).

For profiling - fot start you need to be programmer to analyze results and spot parts which are unusually expensive (for example if "updateAppIcon" takes 40% of CPU then something went wrong, but it is rarely so obvious). Also, you need to configure things for profiling - and I have not done it so far myself.

So, yes - it is not really easy and is basically subset of programming and more advanced one.

**Reply by thany** [reply]
Very subjectively, clearing cache seems to have helped a little bit.

**Comment by Helium314** [comment]
The "old" phone is definitely not a problem, I use a much older S4 mini (plus).

What may help:
* decrease brightness or use the dark theme
* disable auto-upload
* if you know where you will be mapping, download the area manually before you go, so the auto-download does not start
* maybe follow-mode also has some influcence on battery

On my phone, screen and auto-upload are the battery (h)eaters.

**Reply by thany** [reply]
* Dark theme is definitely something I can try, since my phone has an OLED screen, iirc. Good one.
* Decreasing brightness is not really an option for an app that is meant to be used outdoors... I usually need sunglasses as well, so I usually need full blast brightness.
* Surely uploading and downloading doesn't eat that much battery? Isn't it just data? Whatsapp in the background uses data too and hardly any CPU time...
* I already manually move & rotate the map as I see fit. Works better for me anyway.

**Reply by Helium314** [reply]
> Surely uploading and downloading doesn't eat that much battery? Isn't it just data? Whatsapp in the background uses data too and hardly any CPU time...

Data transfers *do* use energy. You can simply give it a try. Also there is some further processing done on each upload, like generating quests.

**Reply by mnalis** [reply]
I agree with @Helium314 estimates. Screen (especially at full brightness) is absolutely the biggest power user for me. It's followed by mobile data, followed by GPS, and only then by CPU (depending on the app)

@thany You can also test how much power just your display uses at full brightness. 

So open some app which doesn't use any other resources but screen (like PDF viewer, or even this github page) and see how much battery it drains in an hour. SC will need to use same amount for light (the biggest use of power, I'd say at least 60% is display). Then compare how much does SC drains the battery with screen on. 

If you see huge differences (e.g. 10% battery used while in PDF viewer, but 60% in SC) then that is something to investigate.
If however the difference is small (e.g. 40% battery used in PDF viewer, but 44% in SC), then it is likely not worth trying to improve SC there (besides trying dark mode) - but shorter screen dim/off timeouts might help if there are periods when screen is on but you're not actively solving quests.

**Reply by thany** [reply]
Yeah that's probably right. Generating light is the one thing phone manufacturers probably can't significantly improve on anymore, as we're approaching the very limits of the tech that's being used (we're nowhere near the theoretical limit yet, i.e. generating light without any heat).

My screen is already OLED, which is more efficient in darkmode than backlit LCD. And darkmode does indeed help battery usage. I do wish darkmode had slightly better contrast, because it also makes it harder to read the map in sunlight.

And I do prefer walking during daylight hours 🙂

**Reply by mnalis** [reply]
Yes, I agree that current battery-backlit-display tech will improve much on power usage...

But we do have [e-ink / e-paper](https://en.wikipedia.org/wiki/Electronic_paper) displays which are super readable in sunlight (the brighter the sun is, the better!) and use **no power** to display (only extremely little power is used to refresh the screen. I use my old Kindle to display text 24/7 for weeks on _single charge_, and that is on more-than-a-decade old dying battery). They have even been used on some [mobile phones](https://en.wikipedia.org/wiki/Electronic_paper#Mobile_phones) including [Android smartphones](https://www.e-ink-info.com/e-ink-devices/mobile-phones) -  I'd love to try one of those to see how it improved. The issue with current state of technology for general use when I last saw them  was with their annoying refreshing of screen (maybe it improved drastically in recent years?) And they needed separate backlight (or externally fitted light) if one wanted to use them in the dark (same as with regular paper book), but that was really nice for me -- much nicer on the eyes then regular backlit OLED/LCD tablets.

Also, old **_non-backlit_** [LCDs](https://en.wikipedia.org/wiki/Liquid-crystal_display) used almost no power and didn't have an e-ink refresh problem; but had somewhat lower contrast, are not readable _at all_ in dark (although there have been backlight-on-demand for them), and usually have glare problem in direct sunlight depending on the angle they're being held. I still use black&white handheld devices with that technology and it works good enough for me under both sun and artificial lightning, while using very little power.

And then there is a [promise](https://www.androidleo.com/samsung-galaxy-transparent-phone-price-specs/) of [transparent display](https://en.wikipedia.org/wiki/See-through_display) which woud use sun itself for backlight, instead of wasting precious battery for backlight while in sunlight.

**Reply by thany** [reply]
> And then there is a [promise](https://www.androidleo.com/samsung-galaxy-transparent-phone-price-specs/) of [transparent display](https://en.wikipedia.org/wiki/See-through_display) which woud use sun itself for backlight

More like frontlight actually, innit 😀

**Comment by westnordost** [comment]
There is something I suspect to eat a lot of battery, and that is the update of the map display each render frame. Usually the display of the map is only updated if you move the map (etc). That it updates every render frame is triggered bythe view that displays the compass direction you hold your phone in (that blue triangle thing). It is not displayed when GPS is off. So to test if this really eats a lot of battery, you can compare

1. battery usage as normal vs
2. battery usage when GPS is off

For an ideal test setup, have your phone on 100%, then go mapping for a few hours or so each, then compare battery % for 1 and 2.
